### ISSUE TITLE
brainstorming

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/BDD#1]
I want to have a better BDD library than the existing ones.

**Comment by inf3rno** [comment]
What I tried so far are the 3 existing cucumber libraries and vanilla mocha. I think cucumber is a dead end, because the syntax of the text is too formal. When somebody wants to write about a feature, then they don't want to think about the syntax. That's why they are not developers... I think I need something a lot closer to real texts. I am not sure whether I'll need NLP. I need to think more about it. I don't like the vanilla mocha approach either, because I cannot reuse fixtures with it. I always need to copy paste or create separate functions for them to reuse. I don't like any of these approaches. I'd rather have something similar to feature files, but I would use MD files and natural text without strict syntax instead. I need to describe a few features and check what I can do. I need an MD parser, but that I can solve. I think I can use the bold and italic styles and the headers to connect it to the test code somehow, but I need to think more about it.

**Comment by inf3rno** [comment]
I might start this project in 2020, I won't have much time for it this year. I don't want it to depend on NLP, at least writing a good NLP solution is extremely hard and it takes too many years to wait for it. I am not telling, that we cannot use existing NLP tools, I am just telling, that it should not depend on non-existing tools I'd need to write.

**Comment by inf3rno** [comment]
It turned out what I am talking about is close to literate programming: https://en.wikipedia.org/wiki/Literate_programming
Even there is an R based solution that uses markdown: http://kbroman.org/knitr_knutshell/pages/Rmarkdown.html

**Comment by inf3rno** [comment]
Actually I could do something like this:

``````
# feature - ...
## scenario - ...
A more detailed description.
```js
function sometest(){
    // ...
}
 ```
Some more description.
```js
function moretest(){
    // ...
}
 ```
blah-blah
We can add an UML diagram too as an attached image.
``````

We need to bundle the tests and make a code map. If we want to make the bundle with browserify, then we'll need a browserify plugin to handle requires in MD files. But we can make the bundle without browserify too, I just don't know if double mapping is possible. It is a lot easier to do it for nodejs only first.

I am not sure if I want to keep the feature and scenario organization. We could use the original non-hierarchical (web) approach of literate programming, so using a graph instead of a tree data structure behind the tests and codes.

For testing I like this approach, for writing code I would not like it because of the extra work.

**Comment by inf3rno** [comment]
I am curious if I can use this literate programming for writing rapidly changing OO code. Some people claim it is very bad for that usage, because just like comments, the texts will rot after time, because it is hard to maintain them and most of the people just neglect them.

**Comment by inf3rno** [comment]
http://ashkenas.com/docco/
I think this explains the best what this literate programming is about. It is more like this than the upper example I had: 


``````
# feature - ...
## scenario - ...
A more detailed description.
```js
function sometest(){
    // ...
```
some examplanation about half of the function body
```js
    // ...
}
 ```
Some more description.
```js
function moretest(){
    // ...
 ```
blah
```js
    // ...
}
 ```
blah-blah
We can add an UML diagram too as an attached image.
``````

**Comment by inf3rno** [comment]
What I don't like about it, that it breaks the structure of the code. That's why I found it better to use function descriptions instead of single line descriptions. If a function is not simple enough we can break it to smaller functions. I think that's how this literate programming should be used.

**Comment by inf3rno** [comment]
let's check an example:

This Todo class is about describing tasks we intend to do. The `text` parameter is the description of the task, the `expiry` parameter is the expiration date we choose for the task.
```js
var Todo = function (text, expiry){
    this.text = text;
    this.expiry = expiry;
}
```
We need to decide if the Task is beyond the deadline. In order to do it we added the `isExpired()` method.
```js
Todo.prototype.isExpired = function (){
    var now = new Date();
    return now > this.expiry;
};
```
and so on ...

**Comment by inf3rno** [comment]
It is good if you want to reason about why you added a certain method, class, etc. But even in this small example there were problems. I almost added `return (new Date()) > this.expiry;` to the code, because I explained it in the text. So the quality of the code would be worse because we explain many things in the text. This leads to the classic math approach where we work with x and y instead of proper expressive variable names. It leads easier to errors, because it is harder to interpret what the actual line does. I think the concept is not bad, but it should be used differently. I mean instead of explaining each variable, they should be self explanatory and all we need to explain is the function name, the parameter list and the return value. Just like by the normal documentation. If we want to explain the algorithm, then it is better to do it with a hyperlink or with some text, that does not refer to any line in the code, but describes the math behind the algorithm. The text will usually contain the variable names we use, but even that is not necessary. So I guess what we describe is the contract and the what it does and maybe a little bit about how it does if it is not evident even with proper variable names. We could even use links to the algorithm description to elaborate about it, instead of adding many pages to the text and explaining it there... I think linking it is a best practice, I mean we just use it, so we should not explain it.

**Comment by inf3rno** [comment]
It isn't that bad, I might write my own transpiler for it and develop this way. Still I cannot spare the brainstorming about the proper model. Implementing it without thinking it through might lead to wasting time on writing bad code. DDD people usually write that it is not a problem, because having a bad model is still better than not having a model. This is good if you are working on a model in a team, but even in that case I could just add another issue to the issue tracker and discuss it, instead of implementing it and refactoring it by each decision. So even with this method I still need to discuss and think through the model first and implement it only after that. What is interesting here, that in theory we can use the text for thinking it through and if the text is final, then we can start to implement it. I am not sure though that in that case the upper text is a good start:

> This Todo class is about describing tasks we intend to do. The text parameter is the description of the task, the expiry parameter is the expiration date we choose for the task.

> We need to decide if the Task is beyond the deadline. In order to do it we added the isExpired() method, which returns that info.

> and so on ...

It might be better to leave out the contract and write about the problem needs to be solved.

> want "describing tasks we intend to do."
> with the "description of the task," 
> and with the "expiration date we choose for the task."

> we should be able to "decide if the Task is beyond the deadline."

> "and so on ..."

**Comment by inf3rno** [comment]
let's write it back to the example:

We want to describe tasks we intend to do with the description text and the expiration date.
```js
var Todo = function (text, expiry){
    this.text = text;
    this.expiry = expiry;
}
```
We should be able to decide if the task is beyond the deadline.
```js
Todo.prototype.isExpired = function (){
    var now = new Date();
    return now > this.expiry;
};
```
and so on ...

**Comment by inf3rno** [comment]
I think that's how it should be used. And it helps a lot by naming the functions and classes. In this case we could use `isBeyondDeadline()` instead of `isExpired()`. It sounds a lot more natural... I think literate programming should work this way instead of explaining the code it should be about what problem the code solves.

**Comment by inf3rno** [comment]
A very interesting question here if we should add the tests to the same file e.g. in the upper case we could write a tests like

it should be return true if the task is beyond deadline
```js
var task = new Task("text", new Date(0));
expect(task.isExpired()).to.equal(true);
```
it should return false if the task is not beyond deadline
```js
var task = new Task("text", new Date(new Date().getTime() + 1000));
expect(task.isExpired()).to.equal(false);
```

**Comment by inf3rno** [comment]
I have to think about it. I mean the texts are evident, still they are needed to know what we are testing. Still we are talking about return values, not about the problem. So either we move that description to the test function name, or we need to abstract out the return value term somehow, which I think is impossible.

I noticed another thing. The *```js* syntax is not good, because we have to write it many times. It is a lot better to use indentation for code blocks. The language will be evident, because we won't use a multilanguage transpiler...

**Comment by inf3rno** [comment]
We want to describe tasks we intend to do with the description text and the expiration date.

```js
	var Todo = function (text, expiry){
		this.text = text;
		this.expiry = expiry;
	}
```

We should be able to decide if the task is beyond the deadline.

```js
	Todo.prototype.isExpired = function (){
		var now = new Date();
		return now > this.expiry;
	};
```

In order to do that we need a function which should return true when the task is beyond deadline, 

```js,test
var task = new Task("text", new Date(0));
expect(task.isExpired()).to.equal(true);
```

and it should return false when the task is not beyond deadline.

```js,test
var task = new Task("text", new Date(new Date().getTime() + 1000));
expect(task.isExpired()).to.equal(false);
```

**Comment by inf3rno** [comment]
I don't feel this right though.

**Comment by inf3rno** [comment]
I think we still got it wrong. I mean programming involves class and instance graphs, not hierarchies. So adding sections, etc for abstraction layers does not help that much, because the whole thing is not that hierarchical. I think a single file or directory should contain an abstraction level, not sections, and we need to use hyperlinks if we want to include lower level classes.

**Comment by inf3rno** [comment]
We could add namespaces for abstraction levels too. I don't think we really need files and directories. From git perspective it is better to have files and directories. From the editor's perspective it does not really matter if a single or multiple files contain the info. It is easier to load, index and cross reference a single file though.

**Comment by inf3rno** [comment]
As of examples I am not convinced that tests are good for that. Well maybe for user manual examples they are good, but for tutorials they are really bad unless we tests several features at once. On the other hand running tutorial examples is required I think. My experience that otherwise they will contain many bugs.

**Comment by inf3rno** [comment]
We want to describe tasks we intend to do with the description text and the expiration date.

```js
var Todo = function (text, expiry){
	this.text = text;
	this.expiry = expiry;
}
```

We should be able to decide if the task is beyond the deadline. In order to do that we need a function, which should return true when the task is beyond deadline.

```js
Todo.prototype.isExpired = function (){
	var now = new Date();
	return now > this.expiry;
};
```
By testing it we first check if it returns true when the task is beyond deadline.

```js,test
var task = new Task("text", new Date(0));
expect(task.isExpired()).to.equal(true);
```

Next we check if it returns false, when the task is not beyond deadline.

```js,test
var task = new Task("text", new Date(new Date().getTime() + 1000));
expect(task.isExpired()).to.equal(false);
```

**Comment by inf3rno** [comment]
Another idea is adding text about the tests in an explicit way. For me this feels a lot better than the previous solution even though it is more words. I am not sure if we want to discuss implementation details in the text. I was against it, but maybe I was wrong. For BDD it is not a good thing, but if we write text for all the classes, then we will need it. I think both styles can work, I like the pure BDD approach better, but the other one might help to name variables and design the model.

**Comment by inf3rno** [comment]
We could even do a hybrid solution and import the source code of the Todo class as a web resource: `/lib/todo.js#Todo`. The same is true for the tests. So they can even be in separate files and imported into the text. The fun part that we can make them editable too. So they can stay in a comment and text-less source file and we can still keep our text. I like this solution better, because it is compatible with Webstorm and other IDE-s and if a developer from the team follows a code first, document later approach, then they can do it with a markdown editor any time.

**Comment by inf3rno** [comment]
Another GUI solution for documenting code: https://underscorejs.org/docs/underscore.html Left side is the text, right side is the code vertically aligned.

**Comment by inf3rno** [comment]
Literate programming follows the order of the thoughts, while the code follows the order the interpreter wants. I think by embedding code we can do both.

**Comment by inf3rno** [comment]
https://github.com/campoy/embedmd
This is something similar I want for the code of the MD file. Regex is just not a good fit for the task. What I really need is a partial or a full AST and find the position on that instead of using regex. For classes it will be `todo.js#Todo.isExpired` for example. Another option to use the index: `todo.js#from=123&to=456`. I like the previous one better, because I don't have to override it every time I change the source code. I don't plan to add per line mapping support. The per method mapping should be enough. If I make up my mind I could use regex or exact match or the line number from the beginning of the block.

**Comment by inf3rno** [comment]
https://hackaday.com/2016/06/06/learn-to-program-with-literate-programming/
Interesting article. According to it literate programming uses macros. Which means we can add an entire method or class into a macro. Those macro calls have really bad syntax, something like `<<<bullshit>>>`. What I'd like is a simple intentation if I want to add code to markdown. If not, then embed with markdown.

**Comment by inf3rno** [comment]
What if we want to do this with a minimalist approach? We can use the wiki (and even version control it) for the documentation. In theory we can add code snippets to markdown. So if the goal is only documenting the code, then it is not a problem to embed it, and we don't need any special tools for that. If we want to do this, then we need to do it in the literate way and use the same for editing the code and maybe even macros. Otherwise working on it does not really worth the effort.

**Comment by inf3rno** [comment]
Another option to use it for tutorials only: http://softwarephilosophy.ninja/literate-programming-really-works-for-tutorials It is a good fit for them.
